File: concurrent-inventory-catalog-jobs-corrupt-stock/python/reconcile_concurrent_inventory_drift.py
"""Detect and repair BigCommerce stock corrupted by concurrent bulk jobs.

BigCommerce's Inventory API processes absolute and relative adjustments
asynchronously through its own internal queue, and its documentation warns
that running Inventory API bulk operations in parallel with Catalog API or
Orders API bulk operations "may cause unpredictable, incorrect calculation
results." Relative adjustments do a read-modify-write against the current
stored total_inventory_onhand, so a catalog bulk edit that also touches
inventory_level, or an order bulk job decrementing stock, can race an
inventory adjustment job on the same SKU and location and silently clobber
or double-apply it. There is also a documented propagation delay between an
adjustment call returning 200 and the new value being reliably readable via
GET, which widens the race window.

BigCommerce does not expose a public adjustment audit-trail endpoint, so
this job reconstructs the expected on-hand for each SKU and location from
the integration's own adjustment ledger, compares it against the actual
total_inventory_onhand BigCommerce reports, and pushes a corrective
absolute adjustment only where the two disagree beyond a tolerance. Every
write is re-verified with a fresh GET before the SKU is marked reconciled.
Run once after any window where inventory and catalog or order bulk jobs
overlapped, then gate all future jobs behind a per-store_hash mutex so this
does not happen again.

Guide: https://www.allanninal.dev/bigcommerce/concurrent-inventory-catalog-jobs-corrupt-stock/
"""
import os
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_concurrent_inventory_drift")

STORE_HASH = os.environ.get("BIGCOMMERCE_STORE_HASH", "example_hash")
ACCESS_TOKEN = os.environ.get("BIGCOMMERCE_ACCESS_TOKEN", "bc_dummy")
API_BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/v3"
STOCK_TOLERANCE = int(os.environ.get("STOCK_TOLERANCE", "0"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

MAX_ITEMS_PER_ADJUSTMENT_CALL = 2000

HEADERS = {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def bc_get(path, params=None):
    r = requests.get(f"{API_BASE}{path}", headers=HEADERS, params=params or {}, timeout=30)
    r.raise_for_status()
    return r.json()


def bc_put(path, body):
    r = requests.put(f"{API_BASE}{path}", headers=HEADERS, json=body, timeout=30)
    r.raise_for_status()
    return r.json()


def is_inventory_corrupted(actual_on_hand: int, expected_on_hand: int, tolerance: int = 0) -> bool:
    """Pure decision. No network, no side effects.

    Returns True (flag for repair) when the actual on-hand BigCommerce
    reports differs from the expected on-hand reconstructed from our own
    adjustment ledger by more than tolerance. Returns False otherwise.
    """
    return abs(actual_on_hand - expected_on_hand) > tolerance


def build_correction_payload(sku: str, location_id: int, expected_on_hand: int) -> dict:
    """Pure payload builder. No network, no side effects.

    Returns the exact item dict the absolute-adjustment request body
    expects for one SKU and location.
    """
    return {"location_id": location_id, "sku": sku, "quantity": expected_on_hand}


def touched_products(job_start_ts):
    """Page through products and variants modified since job_start_ts."""
    page = 1
    while True:
        resp = bc_get(
            "/catalog/products",
            {
                "include": "variants",
                "date_modified:min": job_start_ts,
                "page": page,
                "limit": 250,
            },
        )
        rows = resp.get("data", [])
        if not rows:
            return
        for product in rows:
            yield product
        page += 1


def actual_inventory_items(location_id, skus):
    """Read current total_inventory_onhand for a batch of SKUs at a location."""
    if not skus:
        return []
    resp = bc_get(f"/inventory/locations/{location_id}/items", {"sku__in": ",".join(skus)})
    return resp.get("data", [])


def expected_on_hand_from_ledger(sku, location_id, ledger):
    """Reconstruct expected on-hand from our own adjustment log.

    ledger maps (sku, location_id) -> expected on-hand, built from a
    baseline plus every relative or absolute adjustment this integration
    issued during the overlapping window. BigCommerce does not expose a
    public adjustment audit-trail endpoint, so this ledger has to be kept
    by the integration itself.
    """
    return ledger.get((sku, location_id))


def push_absolute_adjustment(items):
    body = {"reason": "reconciliation-after-concurrent-jobs", "items": items}
    return bc_put("/inventory/adjustments/absolute", body)


def batched(seq, size):
    for i in range(0, len(seq), size):
        yield seq[i : i + size]


def run(job_start_ts=None, ledger=None):
    """ledger: dict mapping (sku, location_id) -> expected on-hand int.

    In production this is built from the integration's own persisted
    adjustment history, not passed in by hand.
    """
    job_start_ts = job_start_ts or os.environ.get("JOB_START_TS")
    ledger = ledger or {}

    if not job_start_ts:
        raise SystemExit(
            "run() needs job_start_ts and a persisted adjustment ledger from your own "
            "integration. Call run(job_start_ts, ledger) from your scheduler."
        )

    flagged = []

    for product in touched_products(job_start_ts):
        variants = product.get("variants") or [{
            "sku": product.get("sku"),
            "inventory_level": product.get("inventory_level"),
        }]
        for variant in variants:
            sku = variant.get("sku")
            if not sku:
                continue
            for location_id in sorted({loc for (s, loc) in ledger if s == sku}):
                items = actual_inventory_items(location_id, [sku])
                for item in items:
                    actual = item.get("total_inventory_onhand")
                    expected = expected_on_hand_from_ledger(sku, location_id, ledger)
                    if actual is None or expected is None:
                        continue
                    if is_inventory_corrupted(actual, expected, STOCK_TOLERANCE):
                        flagged.append((sku, location_id, actual, expected))

    log.info("Found %d SKU/location pair(s) with drift beyond tolerance %d.", len(flagged), STOCK_TOLERANCE)

    corrected = 0
    for batch in batched(flagged, MAX_ITEMS_PER_ADJUSTMENT_CALL):
        payload_items = [
            build_correction_payload(sku, location_id, expected)
            for (sku, location_id, actual, expected) in batch
        ]
        for (sku, location_id, actual, expected) in batch:
            log.info(
                "sku=%s location_id=%s actual_on_hand=%s expected_on_hand=%s (%s)",
                sku, location_id, actual, expected,
                "dry run" if DRY_RUN else "correcting",
            )
        if DRY_RUN:
            continue

        push_absolute_adjustment(payload_items)

        by_location = {}
        for (sku, location_id, _actual, expected) in batch:
            by_location.setdefault(location_id, []).append((sku, expected))

        for location_id, sku_expected in by_location.items():
            skus = [sku for (sku, _e) in sku_expected]
            verify_items = actual_inventory_items(location_id, skus)
            verify_by_sku = {i.get("sku"): i.get("total_inventory_onhand") for i in verify_items}
            for sku, expected in sku_expected:
                if verify_by_sku.get(sku) == expected:
                    corrected += 1
                else:
                    log.warning(
                        "Re-verify failed for sku=%s location_id=%s expected=%s got=%s",
                        sku, location_id, expected, verify_by_sku.get(sku),
                    )

    log.info(
        "Done. %d SKU/location pair(s) %s.",
        len(flagged), "would be corrected" if DRY_RUN else f"corrected ({corrected} re-verified)",
    )


if __name__ == "__main__":
    run()

File: concurrent-inventory-catalog-jobs-corrupt-stock/node/reconcile-concurrent-inventory-drift.js
/**
 * Detect and repair BigCommerce stock corrupted by concurrent bulk jobs.
 *
 * BigCommerce's Inventory API processes absolute and relative adjustments
 * asynchronously through its own internal queue, and its documentation warns
 * that running Inventory API bulk operations in parallel with Catalog API or
 * Orders API bulk operations "may cause unpredictable, incorrect calculation
 * results." Relative adjustments do a read-modify-write against the current
 * stored total_inventory_onhand, so a catalog bulk edit that also touches
 * inventory_level, or an order bulk job decrementing stock, can race an
 * inventory adjustment job on the same SKU and location and silently clobber
 * or double-apply it. There is also a documented propagation delay between an
 * adjustment call returning 200 and the new value being reliably readable via
 * GET, which widens the race window.
 *
 * BigCommerce does not expose a public adjustment audit-trail endpoint, so
 * this job reconstructs the expected on-hand for each SKU and location from
 * the integration's own adjustment ledger, compares it against the actual
 * total_inventory_onhand BigCommerce reports, and pushes a corrective
 * absolute adjustment only where the two disagree beyond a tolerance. Every
 * write is re-verified with a fresh GET before the SKU is marked reconciled.
 * Gate all future inventory and catalog or order bulk jobs behind a single
 * per-store_hash mutex so this does not happen again.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/concurrent-inventory-catalog-jobs-corrupt-stock/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const API_BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/v3`;
const STOCK_TOLERANCE = Number(process.env.STOCK_TOLERANCE || 0);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const MAX_ITEMS_PER_ADJUSTMENT_CALL = 2000;

const HEADERS = {
  "X-Auth-Token": ACCESS_TOKEN,
  "Content-Type": "application/json",
  Accept: "application/json",
};

/**
 * Pure decision. No network, no side effects.
 *
 * Returns true (flag for repair) when the actual on-hand BigCommerce
 * reports differs from the expected on-hand reconstructed from our own
 * adjustment ledger by more than tolerance. Returns false otherwise.
 */
export function isInventoryCorrupted(actualOnHand, expectedOnHand, tolerance = 0) {
  return Math.abs(actualOnHand - expectedOnHand) > tolerance;
}

/**
 * Pure payload builder. No network, no side effects.
 *
 * Returns the exact item object the absolute-adjustment request body
 * expects for one SKU and location.
 */
export function buildCorrectionPayload(sku, locationId, expectedOnHand) {
  return { location_id: locationId, sku, quantity: expectedOnHand };
}

async function bcGet(path, params = {}) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function bcPut(path, body) {
  const res = await fetch(`${API_BASE}${path}`, {
    method: "PUT",
    headers: HEADERS,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function* touchedProducts(jobStartTs) {
  let page = 1;
  while (true) {
    const resp = await bcGet("/catalog/products", {
      include: "variants",
      "date_modified:min": jobStartTs,
      page,
      limit: 250,
    });
    const rows = resp.data || [];
    if (!rows.length) return;
    for (const product of rows) yield product;
    page += 1;
  }
}

async function actualInventoryItems(locationId, skus) {
  if (!skus.length) return [];
  const resp = await bcGet(`/inventory/locations/${locationId}/items`, { sku__in: skus.join(",") });
  return resp.data || [];
}

/**
 * Reconstruct expected on-hand from our own adjustment log.
 *
 * ledger maps a "sku|locationId" key to the expected on-hand, built from a
 * baseline plus every relative or absolute adjustment this integration
 * issued during the overlapping window. BigCommerce does not expose a
 * public adjustment audit-trail endpoint, so this ledger has to be kept by
 * the integration itself.
 */
function expectedOnHandFromLedger(sku, locationId, ledger) {
  return ledger.get(`${sku}|${locationId}`);
}

async function pushAbsoluteAdjustment(items) {
  const body = { reason: "reconciliation-after-concurrent-jobs", items };
  return bcPut("/inventory/adjustments/absolute", body);
}

function* batched(seq, size) {
  for (let i = 0; i < seq.length; i += size) yield seq.slice(i, i + size);
}

/**
 * jobStartTs: ISO timestamp string, only products modified since this are scanned.
 * ledger: Map of "sku|locationId" -> expected on-hand number.
 *
 * In production the ledger is built from the integration's own persisted
 * adjustment history, not passed in by hand.
 */
export async function run(jobStartTs = process.env.JOB_START_TS, ledger = new Map()) {
  if (!jobStartTs) {
    throw new Error(
      "run() needs jobStartTs and a persisted adjustment ledger from your own integration. " +
      "Call run(jobStartTs, ledger) from your scheduler."
    );
  }

  const flagged = [];
  const locationIdsBySku = new Map();
  for (const key of ledger.keys()) {
    const [sku, locationIdStr] = key.split("|");
    const locationId = Number(locationIdStr);
    if (!locationIdsBySku.has(sku)) locationIdsBySku.set(sku, []);
    locationIdsBySku.get(sku).push(locationId);
  }

  for await (const product of touchedProducts(jobStartTs)) {
    const variants = product.variants && product.variants.length
      ? product.variants
      : [{ sku: product.sku, inventory_level: product.inventory_level }];

    for (const variant of variants) {
      const sku = variant.sku;
      if (!sku) continue;
      const locationIds = locationIdsBySku.get(sku) || [];
      for (const locationId of locationIds) {
        const items = await actualInventoryItems(locationId, [sku]);
        for (const item of items) {
          const actual = item.total_inventory_onhand;
          const expected = expectedOnHandFromLedger(sku, locationId, ledger);
          if (actual == null || expected == null) continue;
          if (isInventoryCorrupted(actual, expected, STOCK_TOLERANCE)) {
            flagged.push({ sku, locationId, actual, expected });
          }
        }
      }
    }
  }

  console.log(`Found ${flagged.length} SKU/location pair(s) with drift beyond tolerance ${STOCK_TOLERANCE}.`);

  let corrected = 0;
  for (const batch of batched(flagged, MAX_ITEMS_PER_ADJUSTMENT_CALL)) {
    const payloadItems = batch.map((f) => buildCorrectionPayload(f.sku, f.locationId, f.expected));

    for (const f of batch) {
      console.log(
        `sku=${f.sku} location_id=${f.locationId} actual_on_hand=${f.actual} expected_on_hand=${f.expected} ` +
        `(${DRY_RUN ? "dry run" : "correcting"})`
      );
    }
    if (DRY_RUN) continue;

    await pushAbsoluteAdjustment(payloadItems);

    const byLocation = new Map();
    for (const f of batch) {
      if (!byLocation.has(f.locationId)) byLocation.set(f.locationId, []);
      byLocation.get(f.locationId).push(f);
    }

    for (const [locationId, entries] of byLocation) {
      const skus = entries.map((e) => e.sku);
      const verifyItems = await actualInventoryItems(locationId, skus);
      const verifyBySku = new Map(verifyItems.map((i) => [i.sku, i.total_inventory_onhand]));
      for (const e of entries) {
        if (verifyBySku.get(e.sku) === e.expected) {
          corrected += 1;
        } else {
          console.warn(
            `Re-verify failed for sku=${e.sku} location_id=${locationId} expected=${e.expected} got=${verifyBySku.get(e.sku)}`
          );
        }
      }
    }
  }

  console.log(
    `Done. ${flagged.length} SKU/location pair(s) ${DRY_RUN ? "would be corrected" : `corrected (${corrected} re-verified)`}.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
